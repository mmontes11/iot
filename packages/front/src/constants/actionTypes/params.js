export const PARAM_SELECT = "PARAM_SELECT";
export const PARAM_UPDATE = "PARAM_UPDATE";
export const PARAM_RESET = "PARAM_RESET";
export const PARAM_REQUEST = "PARAM_REQUEST";
export const PARAM_REQUEST_SUCCESS = "PARAM_REQUEST_SUCCESS";
export const PARAM_REQUEST_ERROR = "PARAM_REQUEST_ERROR";
export const PARAM_ITEMS_UPDATED = "PARAM_ITEMS_UPDATED";
export const PARAM_DISABLE = "PARAM_DISABLE";
