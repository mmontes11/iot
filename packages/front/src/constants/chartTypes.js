export const BARCHART = "BARCHART";
export const LINECHART = "LINECHART";
export const REALTIME = "REALTIME";
