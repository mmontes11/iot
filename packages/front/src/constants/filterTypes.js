export const THING_FILTER_TYPE = "thing";
export const DATE_FILTER_TYPE = "date";
export const FILTER_TYPES = [THING_FILTER_TYPE, DATE_FILTER_TYPE];
