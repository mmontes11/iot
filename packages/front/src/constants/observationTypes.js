export const EVENT_TYPE = "event";
export const MEASUREMENT_TYPE = "measurement";
export const OBSERVATION_TYPES = [EVENT_TYPE, MEASUREMENT_TYPE];
