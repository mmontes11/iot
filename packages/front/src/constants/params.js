export const TYPE = "TYPE";
export const OBSERVATION = "OBSERVATION";
export const GROUPBY = "GROUPBY";
export const THING = "THING";
