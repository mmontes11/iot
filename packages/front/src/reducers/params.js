import {
  PARAM_SELECT,
  PARAM_UPDATE,
  PARAM_RESET,
  PARAM_REQUEST,
  PARAM_REQUEST_SUCCESS,
  PARAM_REQUEST_ERROR,
  PARAM_ITEMS_UPDATED,
  PARAM_DISABLE,
} from "constants/actionTypes/params";
import { RESET } from "constants/actionTypes/common";
import deepEqual from "deep-equal";

export const initialState = {
  byId: {},
  allIds: [],
  isResetDisabled: true,
};

const paramInitialState = {
  items: [],
  isActive: false,
  isLoading: false,
  isDisabled: false,
  selectedItem: null,
};

const addParamIfNeeded = (state, paramId) => {
  if (state.allIds.includes(paramId)) {
    return state;
  }
  return {
    ...state,
    byId: {
      ...state.byId,
      [paramId]: paramInitialState,
    },
    allIds: [...state.allIds, paramId],
  };
};

const updateParam = (state, paramId, param) => ({
  ...state,
  byId: {
    ...state.byId,
    [paramId]: param,
  },
});

const allParamsInInitialState = state => Object.values(state.byId).every(param => deepEqual(param, paramInitialState));

export default (state = initialState, { type, param: paramId, selectedItem, items }) => {
  if (type === RESET) {
    return initialState;
  }
  if (!paramId) {
    return state;
  }
  let newState = { ...state };
  newState = addParamIfNeeded(newState, paramId);
  const param = newState.byId[paramId];
  switch (type) {
    case PARAM_SELECT:
      return updateParam(newState, paramId, { ...param, isActive: !param.isActive });
    case PARAM_UPDATE:
      return { ...updateParam(newState, paramId, { ...param, isActive: false, selectedItem }), isResetDisabled: false };
    case PARAM_RESET: {
      const newStateWithResetParam = updateParam(newState, paramId, paramInitialState);
      return { ...newStateWithResetParam, isResetDisabled: allParamsInInitialState(newStateWithResetParam) };
    }
    case PARAM_REQUEST:
      return updateParam(newState, paramId, { ...param, isLoading: true, isDisabled: true });
    case PARAM_REQUEST_SUCCESS:
      return updateParam(newState, paramId, { ...param, isLoading: false, isDisabled: false });
    case PARAM_REQUEST_ERROR:
      return updateParam(newState, paramId, { ...param, isLoading: false, isDisabled: false });
    case PARAM_ITEMS_UPDATED:
      return updateParam(newState, paramId, { ...param, items });
    case PARAM_DISABLE:
      return updateParam(newState, paramId, { ...param, isDisabled: true });
    default:
      return state;
  }
};

export const getParam = (state, paramName) => state.byId[paramName] || {};
