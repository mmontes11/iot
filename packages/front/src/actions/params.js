import {
  PARAM_SELECT,
  PARAM_UPDATE,
  PARAM_RESET,
  PARAM_REQUEST,
  PARAM_REQUEST_SUCCESS,
  PARAM_REQUEST_ERROR,
  PARAM_ITEMS_UPDATED,
  PARAM_DISABLE,
} from "constants/actionTypes/params";
import { EVENT_TYPE, MEASUREMENT_TYPE, OBSERVATION_TYPES } from "constants/observationTypes";
import { TYPE, OBSERVATION, GROUPBY, THING } from "constants/params";
import iotClient from "lib/iotClient";
import * as fromState from "reducers";
import { isEmpty } from "helpers/validation";

const requestObservations = (type, onStart, onSuccess, onError) => {
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getTypes()
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getTypes()
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  }
};

const requestObservationsDispatchingActions = (dispatch, type, cb) => {
  if (!type) {
    cb();
    return;
  }
  requestObservations(
    type,
    () => dispatch({ type: PARAM_REQUEST, param: OBSERVATION }),
    res => {
      dispatch({ type: PARAM_REQUEST_SUCCESS, param: OBSERVATION, statusCode: res.statusCode, error: null });
      dispatch({ type: PARAM_RESET, param: OBSERVATION });
      dispatch({ type: PARAM_ITEMS_UPDATED, param: OBSERVATION, items: res.body.types });
      cb();
    },
    error => {
      dispatch({ type: PARAM_REQUEST_ERROR, param: OBSERVATION, statusCode: error.statusCode, error });
      cb();
    },
  );
};

export const selectType = () => dispatch => {
  dispatch({ type: PARAM_ITEMS_UPDATED, param: TYPE, items: OBSERVATION_TYPES });
  dispatch({ type: PARAM_SELECT, param: TYPE });
};

export const updateType = type => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: TYPE, selectedItem: type });
  requestObservationsDispatchingActions(dispatch, type, () => undefined);
};

export const selectObservation = () => (dispatch, getState) => {
  const state = getState();
  const observation = fromState.getParam(state, OBSERVATION);
  const dispatchSelectObservation = () => dispatch({ type: PARAM_SELECT, param: OBSERVATION });
  if (observation && !isEmpty(observation.items)) {
    dispatchSelectObservation();
  } else {
    const type = fromState.getParam(state, TYPE);
    if (type && type.selectedItem) {
      requestObservationsDispatchingActions(dispatch, type.selectedItem, dispatchSelectObservation);
    }
  }
};

export const updateObservation = observation => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: OBSERVATION, selectedItem: observation });
};

export const selectGroupBy = () => (dispatch, getState) => {
  const state = getState();
  const groupBy = fromState.getParam(state, GROUPBY);
  if (groupBy && !isEmpty(groupBy.items)) {
    dispatch({ type: PARAM_SELECT, param: GROUPBY });
  } else {
    dispatch({ type: PARAM_REQUEST, param: GROUPBY });
    iotClient.timePeriodsService
      .getSupportedTimePeriods()
      .then(res => {
        dispatch({ type: PARAM_REQUEST_SUCCESS, param: GROUPBY, statusCode: res.statusCode, error: null });
        dispatch({ type: PARAM_ITEMS_UPDATED, param: GROUPBY, items: res.body.timePeriods });
        dispatch({ type: PARAM_SELECT, param: GROUPBY });
      })
      .catch(error => {
        dispatch({ type: PARAM_REQUEST_ERROR, param: GROUPBY, statusCode: error.statusCode, error });
      });
  }
};

export const updateGroupBy = groupBy => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: GROUPBY, selectedItem: groupBy });
};

export const updateParams = params => dispatch => {
  const [firstParam, ...rest] = Object.keys(params);
  dispatch({ type: PARAM_UPDATE, param: firstParam, selectedItem: params[firstParam] });
  rest.forEach(key => {
    dispatch({ type: PARAM_DISABLE, param: key });
    dispatch({ type: PARAM_UPDATE, param: key, selectedItem: params[key] });
  });
};

export const selectThing = () => (dispatch, getState) => {
  const thing = fromState.getParam(getState(), THING);
  if (thing && !isEmpty(thing.items)) {
    dispatch({ type: PARAM_SELECT, param: THING });
  } else {
    dispatch({ type: PARAM_REQUEST, param: THING });
    iotClient.thingsService
      .getThings(true)
      .then(res => {
        dispatch({ type: PARAM_REQUEST_SUCCESS, param: THING, statusCode: res.statusCode, error: null });
        dispatch({ type: PARAM_ITEMS_UPDATED, param: THING, items: res.body.things });
        dispatch({ type: PARAM_SELECT, param: THING });
      })
      .catch(error => {
        dispatch({ type: PARAM_REQUEST_ERROR, param: THING, statusCode: error.statusCode, error });
      });
  }
};

export const updateThing = thingName => (dispatch, getState) => {
  dispatch({ type: PARAM_UPDATE, param: THING, selectedItem: thingName });
  const thingParam = fromState.getParam(getState(), THING);
  const thing = thingParam.items.find(item => item.name === thingName);
  dispatch({ type: PARAM_RESET, param: TYPE });
  dispatch({ type: PARAM_ITEMS_UPDATED, param: TYPE, items: thing.supportedObservationTypes.measurement });
};

export const selectMeasurementType = () => dispatch => {
  dispatch({ type: PARAM_SELECT, param: TYPE });
};

export const updateMeasurementType = type => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: TYPE, selectedItem: type });
};
