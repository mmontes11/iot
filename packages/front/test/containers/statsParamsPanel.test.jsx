import React from "react";
import { shallow } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router-dom";
import StatsParamsPanel from "containers/statsParamsPanel";
import { TYPE } from "constants/params";
import { initialState, defaultStore } from "../constants";

describe("containers/statsParamsPanel", () => {
  it("renders in initial state", () => {
    const wrapper = shallow(
      <Provider store={defaultStore}>
        <MemoryRouter initialEntries={["/"]} keyLength={0}>
          <StatsParamsPanel onParamsSelected={() => undefined} onReset={() => undefined} />
        </MemoryRouter>
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("renders when TYPE is selected", () => {
    const state = {
      ...initialState,
      params: {
        byId: {
          TYPE: {
            items: ["event", "measurement"],
            isActive: false,
            isLoading: false,
            isDisabled: false,
            selectedItem: "event",
          },
        },
        allIds: [TYPE],
      },
    };
    const store = configureStore([thunk])(state);
    const wrapper = shallow(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/"]} keyLength={0}>
          <StatsParamsPanel onParamsSelected={() => undefined} onReset={() => undefined} />
        </MemoryRouter>
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("renders when loading TYPE", () => {
    const state = {
      ...initialState,
      params: {
        byId: {
          TYPE: {
            items: ["event", "measurement"],
            isActive: false,
            isLoading: true,
            isDisabled: false,
            selectedItem: null,
          },
        },
        allIds: [TYPE],
      },
    };
    const store = configureStore([thunk])(state);
    const wrapper = shallow(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/"]} keyLength={0}>
          <StatsParamsPanel onParamsSelected={() => undefined} onReset={() => undefined} />
        </MemoryRouter>
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
