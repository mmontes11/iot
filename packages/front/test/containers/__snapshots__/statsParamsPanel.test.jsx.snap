// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/statsParamsPanel renders in initial state 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  keyLength={0}
>
  <Connect(StatsParamsPanel)
    onParamsSelected={[Function]}
    onReset={[Function]}
  />
</MemoryRouter>
`;

exports[`containers/statsParamsPanel renders when TYPE is selected 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  keyLength={0}
>
  <Connect(StatsParamsPanel)
    onParamsSelected={[Function]}
    onReset={[Function]}
  />
</MemoryRouter>
`;

exports[`containers/statsParamsPanel renders when loading TYPE 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  keyLength={0}
>
  <Connect(StatsParamsPanel)
    onParamsSelected={[Function]}
    onReset={[Function]}
  />
</MemoryRouter>
`;
